Docker-build fails to detect podman

### Describe the bug

I have an application, which uses `quarkus-container-image-docker` extension and I am trying to build the app on a machine with podman. It works for released version of Quarkus, but fails for SNAPSHOT (starting from the 1st of June 2024)

### Expected behavior

Build succeeds

### Actual behavior

```
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:build (default) on project code-with-quarkus: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.container.image.docker.deployment.DockerProcessor#dockerBuildFromJar threw an exception: java.lang.IllegalStateException: No container runtime was found. Make sure you have either Docker or Podman installed in your environment.
[ERROR] 	at io.quarkus.deployment.util.ContainerRuntimeUtil.detectContainerRuntime(ContainerRuntimeUtil.java:68)
[ERROR] 	at io.quarkus.deployment.util.ContainerRuntimeUtil.detectContainerRuntime(ContainerRuntimeUtil.java:50)
[ERROR] 	at io.quarkus.container.image.docker.common.deployment.CommonProcessor.lambda$getExecutableName$9(CommonProcessor.java:262)
[ERROR] 	at java.base/java.util.Optional.orElseGet(Optional.java:364)
[ERROR] 	at io.quarkus.container.image.docker.common.deployment.CommonProcessor.getExecutableName(CommonProcessor.java:262)
[ERROR] 	at io.quarkus.container.image.docker.common.deployment.CommonProcessor.buildFromJar(CommonProcessor.java:87)
[ERROR] 	at io.quarkus.container.image.docker.deployment.DockerProcessor.dockerBuildFromJar(DockerProcessor.java:62)
[ERROR] 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:483)
[ERROR] -> [Help 1]

```

### How to Reproduce?

1. `wget -O app.zip "https://code.quarkus.io/d?e=rest&e=container-image-docker&cn=code.quarkus.io"`
2. `unzip -q app.zip && cd code-with-quarkus/`
3. This succeeds: ` mvn clean verify -Dquarkus.platform.version=3.11.0 -Dquarkus.container-image.build=true`
4.  This fails:  `mvn clean verify -Dquarkus.platform.group-id=io.quarkus -Dquarkus.platform.version=999-SNAPSHOT -Dquarkus.container-image.build=true`

### Output of `uname -a` or `ver`

6.7.3-200.fc39.x86_64

### Output of `java -version`

Java version: 21.0.1, vendor: Eclipse Adoptium

### Quarkus version or git rev

6bc64a00d8748e56db2e4bb7a49f13f1c30d51de  (the earliest affected build)

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)

### Additional information

```
$ docker --version
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
podman version 5.0.3
```

`mvn clean verify -Dquarkus.platform.version=3.11.0 -Dquarkus.container-image.build=true` actually fails for me...

```
╭─ ~/Desktop/test/code-with-quarkus ············································································· ✔  system Node 
╰─ ./mvnw clean verify -Dquarkus.platform.version=3.11.0 -Dquarkus.container-image.build=true
[INFO] Scanning for projects...
[INFO] 
[INFO] ---------------------< org.acme:code-with-quarkus >---------------------
[INFO] Building code-with-quarkus 1.0.0-SNAPSHOT
[INFO]   from pom.xml
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- clean:3.2.0:clean (default-clean) @ code-with-quarkus ---
[INFO] 
[INFO] --- resources:3.3.1:resources (default-resources) @ code-with-quarkus ---
[INFO] Copying 1 resource from src/main/resources to target/classes
[INFO] 
[INFO] --- quarkus:3.11.0:generate-code (default) @ code-with-quarkus ---
Downloading from central: https://repo1.maven.org/maven2/org/acme/code-with-quarkus/1.0.0-SNAPSHOT/maven-metadata.xml
[INFO] 
[INFO] --- compiler:3.12.1:compile (default-compile) @ code-with-quarkus ---
[INFO] Recompiling the module because of changed source code.
[INFO] Compiling 1 source file with javac [debug release 21] to target/classes
[INFO] 
[INFO] --- quarkus:3.11.0:generate-code-tests (default) @ code-with-quarkus ---
[INFO] 
[INFO] --- resources:3.3.1:testResources (default-testResources) @ code-with-quarkus ---
[INFO] skip non existing resourceDirectory /Users/edeandre/Desktop/test/code-with-quarkus/src/test/resources
[INFO] 
[INFO] --- compiler:3.12.1:testCompile (default-testCompile) @ code-with-quarkus ---
[INFO] Recompiling the module because of changed dependency.
[INFO] Compiling 2 source files with javac [debug release 21] to target/test-classes
[INFO] Annotation processing is enabled because one or more processors were found
  on the class path. A future release of javac may disable annotation processing
  unless at least one processor is specified by name (-processor), or a search
  path is specified (--processor-path, --processor-module-path), or annotation
  processing is enabled explicitly (-proc:only, -proc:full).
  Use -Xlint:-options to suppress this message.
  Use -proc:none to disable annotation processing.
[INFO] 
[INFO] --- surefire:3.2.5:test (default-test) @ code-with-quarkus ---
[INFO] Using auto detected provider org.apache.maven.surefire.junitplatform.JUnitPlatformProvider
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.acme.GreetingResourceTest
2024-06-10 07:57:59,128 INFO  [io.quarkus] (main) code-with-quarkus 1.0.0-SNAPSHOT on JVM (powered by Quarkus 3.11.0) started in 0.987s. Listening on: http://localhost:8081
2024-06-10 07:57:59,129 INFO  [io.quarkus] (main) Profile test activated. 
2024-06-10 07:57:59,130 INFO  [io.quarkus] (main) Installed features: [cdi, rest, smallrye-context-propagation, vertx]
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.667 s -- in org.acme.GreetingResourceTest
2024-06-10 07:57:59,678 INFO  [io.quarkus] (main) code-with-quarkus stopped in 0.010s
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- jar:3.3.0:jar (default-jar) @ code-with-quarkus ---
[INFO] Building jar: /Users/edeandre/Desktop/test/code-with-quarkus/target/code-with-quarkus-1.0.0-SNAPSHOT.jar
[INFO] 
[INFO] --- quarkus:3.11.0:build (default) @ code-with-quarkus ---
[WARNING] [io.quarkus.deployment.util.ContainerRuntimeUtil] Command "docker info" exited with error code 1. Rootless container runtime detection might not be reliable or the container service is not running at all.
[INFO] [io.quarkus.container.image.docker.deployment.DockerProcessor] Starting (local) container image build for jar using docker.
[INFO] [io.quarkus.container.image.docker.deployment.DockerProcessor] Executing the following command to build docker image: 'docker build -f /Users/edeandre/Desktop/test/code-with-quarkus/src/main/docker/Dockerfile.jvm -t edeandre/code-with-quarkus:1.0.0-SNAPSHOT /Users/edeandre/Desktop/test/code-with-quarkus'
[INFO] [io.quarkus.deployment.util.ExecUtil] ERROR: Cannot connect to the Docker daemon at unix:///Users/edeandre/.docker/run/docker.sock. Is the docker daemon running?
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  7.188 s
[INFO] Finished at: 2024-06-10T07:58:01-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.11.0:build (default) on project code-with-quarkus: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.container.image.docker.deployment.DockerProcessor#dockerBuildFromJar threw an exception: java.lang.RuntimeException: Execution of 'docker build -f /Users/edeandre/Desktop/test/code-with-quarkus/src/main/docker/Dockerfile.jvm -t edeandre/code-with-quarkus:1.0.0-SNAPSHOT /Users/edeandre/Desktop/test/code-with-quarkus' failed. See docker output for more details
[ERROR] 	at io.quarkus.container.image.docker.deployment.DockerProcessor.dockerException(DockerProcessor.java:339)
[ERROR] 	at io.quarkus.container.image.docker.deployment.DockerProcessor.createContainerImage(DockerProcessor.java:198)
[ERROR] 	at io.quarkus.container.image.docker.deployment.DockerProcessor.dockerBuildFromJar(DockerProcessor.java:103)
[ERROR] 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[ERROR] 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:849)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:1583)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:483)
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

--- a/core/deployment/src/main/java/io/quarkus/deployment/IsDockerWorking.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/IsDockerWorking.java
@@ -6,6 +6,7 @@ import static io.quarkus.deployment.util.ContainerRuntimeUtil.ContainerRuntime.U
 import java.util.List;
 
 import io.quarkus.deployment.util.ContainerRuntimeUtil;
+import io.quarkus.deployment.util.ContainerRuntimeUtil.ContainerRuntime;
 
 public class IsDockerWorking extends IsContainerRuntimeWorking {
     public IsDockerWorking() {
@@ -19,7 +20,8 @@ public class IsDockerWorking extends IsContainerRuntimeWorking {
     private static class DockerBinaryStrategy implements Strategy {
         @Override
         public Result get() {
-            if (ContainerRuntimeUtil.detectContainerRuntime(false) != UNAVAILABLE) {
+            if (ContainerRuntimeUtil.detectContainerRuntime(false,
+                    ContainerRuntime.DOCKER, ContainerRuntime.PODMAN) != UNAVAILABLE) {
                 return Result.AVAILABLE;
             } else {
                 return Result.UNKNOWN;

--- a/core/deployment/src/main/java/io/quarkus/deployment/IsPodmanWorking.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/IsPodmanWorking.java
@@ -5,6 +5,7 @@ import static io.quarkus.deployment.util.ContainerRuntimeUtil.ContainerRuntime.U
 import java.util.List;
 
 import io.quarkus.deployment.util.ContainerRuntimeUtil;
+import io.quarkus.deployment.util.ContainerRuntimeUtil.ContainerRuntime;
 
 public class IsPodmanWorking extends IsContainerRuntimeWorking {
     public IsPodmanWorking() {
@@ -21,7 +22,11 @@ public class IsPodmanWorking extends IsContainerRuntimeWorking {
     private static class PodmanBinaryStrategy implements Strategy {
         @Override
         public Result get() {
-            return (ContainerRuntimeUtil.detectContainerRuntime(false) != UNAVAILABLE) ? Result.AVAILABLE : Result.UNKNOWN;
+            if (ContainerRuntimeUtil.detectContainerRuntime(false, ContainerRuntime.PODMAN) != UNAVAILABLE) {
+                return Result.AVAILABLE;
+            } else {
+                return Result.UNKNOWN;
+            }
         }
     }
 }

--- a/core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/util/ContainerRuntimeUtil.java
@@ -6,6 +6,7 @@ import java.io.InputStream;
 import java.io.InputStreamReader;
 import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import java.util.concurrent.TimeUnit;
 import java.util.function.Predicate;
@@ -42,21 +43,20 @@ public final class ContainerRuntimeUtil {
      * @return a fully resolved {@link ContainerRuntime} indicating if Docker or Podman is available and in rootless mode or not
      * @throws IllegalStateException if no container runtime was found to build the image
      */
-    public static ContainerRuntime detectContainerRuntime() {
-        return detectContainerRuntime(true);
-    }
-
-    public static ContainerRuntime detectContainerRuntime(List<ContainerRuntime> orderToCheckRuntimes) {
+    public static ContainerRuntime detectContainerRuntime(ContainerRuntime... orderToCheckRuntimes) {
         return detectContainerRuntime(true, orderToCheckRuntimes);
     }
 
-    public static ContainerRuntime detectContainerRuntime(boolean required) {
-        return detectContainerRuntime(required, List.of(ContainerRuntime.DOCKER, ContainerRuntime.PODMAN));
+    public static ContainerRuntime detectContainerRuntime(boolean required, ContainerRuntime... orderToCheckRuntimes) {
+        return detectContainerRuntime(
+                required,
+                ((orderToCheckRuntimes != null) && (orderToCheckRuntimes.length > 0)) ? Arrays.asList(orderToCheckRuntimes)
+                        : List.of(ContainerRuntime.DOCKER, ContainerRuntime.PODMAN));
     }
 
     public static ContainerRuntime detectContainerRuntime(boolean required, List<ContainerRuntime> orderToCheckRuntimes) {
         ContainerRuntime containerRuntime = loadContainerRuntimeFromSystemProperty();
-        if (containerRuntime != null) {
+        if ((containerRuntime != null) && orderToCheckRuntimes.contains(containerRuntime)) {
             return containerRuntime;
         }
 

--- a/extensions/container-image/container-image-docker-common/deployment/src/main/java/io/quarkus/container/image/docker/common/deployment/CommonProcessor.java
+++ b/extensions/container-image/container-image-docker-common/deployment/src/main/java/io/quarkus/container/image/docker/common/deployment/CommonProcessor.java
@@ -56,7 +56,7 @@ public abstract class CommonProcessor<C extends CommonConfig> {
             BuildProducer<ArtifactResultBuildItem> artifactResultProducer,
             BuildProducer<ContainerImageBuilderBuildItem> containerImageBuilder,
             PackageConfig packageConfig,
-            ContainerRuntime containerRuntime) {
+            ContainerRuntime... containerRuntimes) {
 
         var buildContainerImage = buildContainerImageNeeded(containerImageConfig, buildRequest);
         var pushContainerImage = pushContainerImageNeeded(containerImageConfig, pushRequest);
@@ -84,7 +84,7 @@ public abstract class CommonProcessor<C extends CommonConfig> {
                 LOGGER.infof("Starting (local) container image build for jar using %s", getProcessorImplementation());
             }
 
-            var executableName = getExecutableName(config, containerRuntime);
+            var executableName = getExecutableName(config, containerRuntimes);
             var builtContainerImage = createContainerImage(containerImageConfig, config, containerImageInfo, out,
                     dockerfilePaths, buildContainerImage, pushContainerImage, packageConfig, executableName);
 
@@ -113,7 +113,7 @@ public abstract class CommonProcessor<C extends CommonConfig> {
             BuildProducer<ContainerImageBuilderBuildItem> containerImageBuilder,
             PackageConfig packageConfig,
             NativeImageBuildItem nativeImage,
-            ContainerRuntime containerRuntime) {
+            ContainerRuntime... containerRuntimes) {
 
         var buildContainerImage = buildContainerImageNeeded(containerImageConfig, buildRequest);
         var pushContainerImage = pushContainerImageNeeded(containerImageConfig, pushRequest);
@@ -134,7 +134,7 @@ public abstract class CommonProcessor<C extends CommonConfig> {
                 LOGGER.infof("Starting (local) container image build for jar using %s", getProcessorImplementation());
             }
 
-            var executableName = getExecutableName(config, containerRuntime);
+            var executableName = getExecutableName(config, containerRuntimes);
             var dockerfilePaths = getDockerfilePaths(config, true, packageConfig, out);
             var builtContainerImage = createContainerImage(containerImageConfig, config, containerImage, out, dockerfilePaths,
                     buildContainerImage, pushContainerImage, packageConfig, executableName);
@@ -257,9 +257,9 @@ public abstract class CommonProcessor<C extends CommonConfig> {
                                 getProcessorImplementation()));
     }
 
-    private String getExecutableName(C config, ContainerRuntime containerRuntime) {
+    protected String getExecutableName(C config, ContainerRuntime... containerRuntimes) {
         return config.executableName()
-                .orElseGet(() -> detectContainerRuntime(List.of(containerRuntime)).getExecutableName());
+                .orElseGet(() -> detectContainerRuntime(containerRuntimes).getExecutableName());
     }
 
     private DockerfilePaths getDockerfilePaths(C config,

--- a/extensions/container-image/container-image-docker/deployment/src/main/java/io/quarkus/container/image/docker/deployment/DockerProcessor.java
+++ b/extensions/container-image/container-image-docker/deployment/src/main/java/io/quarkus/container/image/docker/deployment/DockerProcessor.java
@@ -61,7 +61,7 @@ public class DockerProcessor extends CommonProcessor<DockerConfig> {
 
         buildFromJar(dockerConfig, dockerStatusBuildItem, containerImageConfig, out, containerImageInfo,
                 buildRequest, pushRequest, artifactResultProducer, containerImageBuilder, packageConfig,
-                ContainerRuntime.DOCKER);
+                ContainerRuntime.DOCKER, ContainerRuntime.PODMAN);
     }
 
     @BuildStep(onlyIf = { IsNormalNotRemoteDev.class, NativeBuild.class, DockerBuild.class })
@@ -81,7 +81,7 @@ public class DockerProcessor extends CommonProcessor<DockerConfig> {
 
         buildFromNativeImage(dockerConfig, dockerStatusBuildItem, containerImageConfig, containerImage,
                 buildRequest, pushRequest, out, artifactResultProducer, containerImageBuilder, packageConfig, nativeImage,
-                ContainerRuntime.DOCKER);
+                ContainerRuntime.DOCKER, ContainerRuntime.PODMAN);
     }
 
     @Override
@@ -145,6 +145,19 @@ public class DockerProcessor extends CommonProcessor<DockerConfig> {
         }
 
         return containerImageInfo.getImage();
+    }
+
+    @Override
+    protected String getExecutableName(DockerConfig config, ContainerRuntime... containerRuntimes) {
+        var executableName = super.getExecutableName(config, containerRuntimes);
+
+        if (!DOCKER.equals(executableName)) {
+            LOG.warnf(
+                    "Using executable %s within the quarkus-container-image-%s extension. Maybe you should use the quarkus-container-image-%s extension instead?",
+                    executableName, DOCKER, executableName);
+        }
+
+        return executableName;
     }
 
     private String[] getDockerBuildArgs(String image,

